src/osx/cocoa/dnd.mm

```diff
@@ -29,6 +31,215 @@
+{
+public:
+    wxOSXPasteboardSourceItem(NSPasteboardItem* item, NSPasteboard* board): m_item(item), m_pasteboard(board)
+    {
+    }
+
+    virtual wxDataFormat::NativeFormat AvailableType(CFArrayRef types) const
+    {
+        return (wxDataFormat::NativeFormat)[m_item availableTypeFromArray:(NSArray*)types];
+    }
+
+    virtual bool GetData( const wxDataFormat& dataFormat, wxMemoryBuffer& target)
+    {
+        return GetData(dataFormat.GetFormatId(), target);
+    }
+
+
+    virtual bool GetData( wxDataFormat::NativeFormat type, wxMemoryBuffer& target)
+    {
+        bool success = false;
+
+        target.Clear();
+
+        CFDataRef flavorData = DoGetData( type );
+
+        if ( flavorData )
+        {
+            CFIndex flavorDataSize = CFDataGetLength(flavorData);
+            size_t allocatedSize = flavorDataSize+4;
+            void * buf = target.GetWriteBuf(allocatedSize);
+            if ( buf )
+            {
+                memset(buf, 0, allocatedSize);
+                memcpy(buf, CFDataGetBytePtr(flavorData), flavorDataSize);
+                target.UngetWriteBuf(flavorDataSize);
+                success = true;
+            }
+        }
+        return success;
+    }
+
+    virtual CFDataRef DoGetData(wxDataFormat::NativeFormat type) const
+    {
+        // before a file promise can be resolved, we must pass a paste location
+        if ( UTTypeConformsTo((CFStringRef)type, kPasteboardTypeFileURLPromise ) )
+        {
+            wxString tempdir = wxFileName::GetTempDir() + wxFILE_SEP_PATH + "wxtemp.XXXXXX";
+            char* result = mkdtemp((char*)tempdir.fn_str().data());
+
+            wxCFRef<CFURLRef> dest(CFURLCreateFromFileSystemRepresentation(NULL, (const UInt8*)result, strlen(result), true));
+            PasteboardRef pboardRef = NULL;
+            PasteboardCreate((CFStringRef)[m_pasteboard name], &pboardRef);
+            if (pboardRef != NULL) {
+                PasteboardSynchronize(pboardRef);
+                PasteboardSetPasteLocation(pboardRef, (CFURLRef)dest);
+                CFRelease(pboardRef);
+            }
+        }
+
+        return (CFDataRef) [m_item dataForType:(NSString*) type];
+    }
+
+    NSPasteboardItem* GetNative() { return m_item; };
+private:
+    NSPasteboardItem* m_item;
+    NSPasteboard* m_pasteboard;
+};
+
+wxOSXPasteboard::wxOSXPasteboard(OSXPasteboard native)
+{
+    m_pasteboard = native;
+}
+
+wxOSXPasteboard::~wxOSXPasteboard()
+{
+    DeleteSinkItems();
+}
+
+void wxOSXPasteboard::DeleteSinkItems()
+{
+    for ( wxVector<wxOSXDataSinkItem*>::iterator it = m_sinkItems.begin();
+     it != m_sinkItems.end();
+     ++it)
+    {
+        delete (*it);
+    }
+    m_sinkItems.clear();
+}
+
+// data sink methods
+
+void wxOSXPasteboard::Clear()
+{
+    [m_pasteboard clearContents];
+    DeleteSinkItems();    
+}
+
+void wxOSXPasteboard::Flush()
+{
+    NSMutableArray* nsarray = [[NSMutableArray alloc] init];
+    for( auto i : m_sinkItems )
```

Comment:
I don't know if we can afford to unconditionally use C++11 in wxOSX now? If yes, we should use range for loops in `DeleteSinkItems()` too.

Reply:
Good question, I'll change this back, but we should research on C++11 in wxOSX, I'd think these are supported by the Xcode versions we support by OS version. But I'll check